Replace Yuni::IO with std functions

payetvin opened the pull request:
The goal of this PR is to remove Yuni::IO functions in important parts of the code (libs and solver)
`std::filesystem::path` is used instead of Yuni strings when needed, more work will need to be done to use `path` on every level

No changes were made in:
- GUI
- deprecated tools
- functions that save the study

payetvin committed:
use path for sts

payetvin committed:
use path for links

payetvin committed:
Revert "use path for sts"

This reverts commit 7873a59e644cf9499c8bfe287f95e53972492c41.

payetvin committed:
fix include

payetvin committed:
remove Yuni::exists in study and scenario

payetvin committed:
remove c_str

payetvin committed:
Yuni IO in checkAndSetStudyFolder

payetvin committed:
fix yuni string

payetvin committed:
io in tryToFindTheVersion

payetvin committed:
io in readVersionFromFile

payetvin committed:
fix yuni convesion in scenario sets

payetvin committed:
io in study internalLoadFromFolder

payetvin committed:
sts loading

payetvin committed:
expliclit string conversion for relocate func

payetvin committed:
importLogsToOutputFolder

payetvin committed:
batchrun exists

payetvin committed:
fix lexically normal

payetvin committed:
fix str

payetvin committed:
add todo

payetvin committed:
Merge branch 'develop' into fix/yuni-fs

payetvin committed:
exist in study output

payetvin committed:
use namespace fs

payetvin committed:
fix compile

payetvin committed:
set template

payetvin committed:
eco

payetvin committed:
fix compile by removing a template param for sets

payetvin committed:
fix ui code

payetvin committed:
hydro allocation

guilpier-code commented on src/tools/batchrun/main.cpp:
```diff
@@ -218,10 +218,10 @@ int main(int argc, char* argv[])
-            IO::MakeAbsolute(tmp, *optSolver);
-            IO::Normalize(solver, tmp);
-            if (not IO::File::Exists(solver))
+            std::string tmp = *optSolver;
```
Do we need **tmp** here ?

guilpier-code commented on src/tools/batchrun/main.cpp:
```diff
@@ -218,10 +218,10 @@ int main(int argc, char* argv[])
-            IO::Normalize(solver, tmp);
-            if (not IO::File::Exists(solver))
+            std::string tmp = *optSolver;
+            std::filesystem::path solverPath = std::filesystem::absolute(tmp);
```
Could we not have directly : 
```C++
std::filesystem::path solverPath = 
        std::filesystem::absolute(optSolver.to<std::string>()).lexically_normal();
if (! std::filesystem::exists(solverPath))
{
    ...
}
```
?

guilpier-code commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -428,17 +428,16 @@ void logLinkDataCheckErrorDirectIndirect(const AreaLink& link,
 }
 } // anonymous namespace
 
-bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyString& folder)
```
Having : 
```C++
namespace fs = std::filesystem;
```
here (as it's done somewhere else in this PR) would good too, I think.

guilpier-code commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -36,6 +36,8 @@
+
 namespace Antares::Data
 {
 namespace // anonymous
```
About function : 
```C++
template<class StringT>
static bool AreaListLoadFromFolderSingleArea(Study& study,
                                             AreaList* list,
                                             Area& area,
                                             StringT& buffer,
                                             const StudyLoadOptions& options)
```

Parameter **buffer** is actually a local variable : we don't need to pass it to the function.
So, we should : 
- remove the **buffer** parameter and the associated template.
- make buffer a local variable.

Besides, I wonder if we could make buffer a **std::filesystem::path**.
Indeed : 
- it can easily turned into a **std::string**, vie **string()** member function.
- Locally (to this function), buffer very is often passed as a **AnyString**, which has a constructor taking a **std::string** as argument.

So every time we pass **buffer**, we could pass **buffer.string()** instead.
Is this possible ?

guilpier-code replied:
Should be done one day.

guilpier-code commented on src/libs/antares/study/importlogs.cpp:
```diff
@@ -36,9 +36,9 @@ void Study::importLogsToOutputFolder(Solver::IResultWriter& resultWriter) const
-    std::string logPath("simulation.log");
-    String from;
-    IO::Normalize(from, logs.logfile());
+    std::filesystem::path logPath = "simulation.log";
```
```C++
std::filesystem::path logPath = "simulation.log";
```
is not a log path, it's rather a filename. 
**logPath**  == > **logFileName** ?

guilpier-code replied:
And I'am not even sure that we need **logPath** : 
```C++
resultWriter.addEntryFromFile("simulation.log", from.string());
```
should work

guilpier-code commented on src/libs/antares/study/parts/short-term-storage/container.cpp:
```diff
@@ -21,7 +21,6 @@
 
 #include <antares/logs/logs.h>
 #include <yuni/io/file.h>
```
**Question** : 

if we replace : 
```C++
Yuni::IO::Directory::Create(folder);
```
with : 
```C++
std::filesystem::create_directories(folder);
```
, could we then remove the `#include <yuni/io/file.h>` ?

guilpier-code commented on src/libs/antares/study/parts/thermal/ecoInput.cpp:
```diff
@@ -58,27 +58,27 @@ bool EconomicInputData::saveToFolder(const AnyString& folder) const
 
-        filename << folder << SEP << "fuelCost.txt";
-        if (IO::File::Exists(filename))
+        filename = fs::path(folder) / "fuelCost.txt";
```
```C++
fs::path filename = = fs::path(folder) / "fuelCost.txt";
```
instead of 2 lines

payetvin committed:
remove useless dir functions, move code in prepareDirectoryHierarchy

payetvin committed:
remove filename

payetvin committed:
remove notice

payetvin committed:
guillaume comments 1

payetvin committed:
simplify simulation.log writing

payetvin committed:
return true if exist

payetvin committed:
use path more

payetvin committed:
use path for addEntryFromFile

payetvin committed:
use std copy

payetvin committed:
if dir create ok

payetvin committed:
try fixing windows compile

payetvin committed:
use readfile for zip writer

payetvin committed:
try fix windows

payetvin committed:
remove yuni in memory writer

payetvin committed:
log folder and file creation

sylvlecl committed:
Merge remote-tracking branch 'origin/develop' into fix/yuni-fs

sylvlecl committed:
Fix merge

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Apply format

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Fix double root in immediate file writer

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Fix use of possible dangling pointer

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

sylvlecl committed:
Fix directory path for windows (#2084)

**Analysis**

On windows:
```cpp
// this returns true as expected
fs::create_directories("/path/to/my/new/directory");

// thisreturns false because of trailing slash ???
fs::create_directories("/path/to/my/new/directory/");
```

It looks more or less like a bug of the STL ... the last slash is
considered as "not created" which would explain it returns false ...

**The fix**

Use `parent_path` instead of `remove_filename`, to get the correct path
of parent directory without the trailing slash.

Signed-off-by: Sylvain Leclerc <sylvain.leclerc@rte-france.com>

payetvin committed:
remove yuni stream in options

payetvin committed:
use getpid

payetvin committed:
remove include

payetvin committed:
remove commented code

payetvin committed:
stream in area list

payetvin committed:
Merge branch 'develop' into fix/yuni-fs

payetvin committed:
Merge branch 'develop' into fix/yuni-fs

sylvlecl commented on src/libs/antares/writer/immediate_file_writer.cpp:
```diff
@@ -83,62 +39,67 @@ ImmediateFileResultWriter::ImmediateFileResultWriter(const char* folderOutput):
 
-    switch (Yuni::IO::File::Copy(filePath.c_str(), fullPath))
+    std::error_code ec;
+    fs::copy(filePath, fullPath, ec);
```
I think it would be simpler to use the message provided by the STL:
```cpp
try {
    fs::copy(filePath, fullPath);
} catch (const fs::filesystem_error& exc) {
    logs.error() << exc.what();
}
```

For example it provides such message:
```
filesystem error: cannot copy: No such file or directory [source] [target]
```
which is richer than just the error code message "No such file or directory".

flomnes commented on src/libs/antares/study/parts/short-term-storage/container.cpp:
```diff
@@ -54,7 +56,7 @@ bool STStorageInput::createSTStorageClustersFromIniFile(const std::string& path)
     }
 
-    logs.debug() << "  :: loading `" << pathIni << "`";
+    logs.debug() << "  :: loading `" << pathIni.string() << "`";
```
Add overloads to Yuni's logs on type `fs::path` ?

sylvlecl commented on src/libs/antares/writer/in_memory_writer.cpp:
```diff
@@ -77,27 +84,30 @@ void InMemoryWriter::addEntryFromBuffer(const std::string& entryPath, std::strin
 }
 
-void InMemoryWriter::addEntryFromFile(const std::string& entryPath, const std::string& filePath)
+static std::string readFile(const fs::path& filePath)
```
For this kind of very generic function (reading a file to a string), we could share it in a technical lib, for example we already have an "IO" library where it could go.

This function would throw an exception on failure, which we could just let through here.

sylvlecl commented on src/libs/antares/writer/zip_writer.cpp:
```diff
@@ -157,29 +159,29 @@ void ZipWriter::addEntryFromBuffer(const std::string& entryPath, std::string& en
 }
 
-void ZipWriter::addEntryFromFile(const std::string& entryPath, const std::string& filePath)
+static std::string readFile(const fs::path& filePath)
```
This could be removed from here by sharing the function in IO lib

sylvlecl reviewed (commented):
A few technical suggestion: in particular share implementation of reading file to string in IO lib

flomnes commented on src/libs/antares/study/importlogs.cpp:
```diff
@@ -51,12 +50,12 @@ void Study::importLogsToOutputFolder(Solver::IResultWriter& resultWriter) const
     }
 
-    resultWriter.addEntryFromFile(logPath, from.c_str());
+    resultWriter.addEntryFromFile("simulation.log", from.string());
```
Not sure about this. "simulation.log" refers to $PWD/simulation.log, which is probably not what we want

payetvin replied:
I don't think this change the previous behavior, we just avoid using a single use var

flomnes commented on src/libs/antares/study/header.cpp:
```diff
@@ -205,27 +205,25 @@ bool StudyHeader::saveToFile(const AnyString& filename, bool upgradeVersion)
-    String abspath, directory;
-    IO::MakeAbsolute(abspath, folder);
-    IO::Normalize(directory, abspath);
+    // folder normalization
+    fs::path abspath = fs::absolute(folder);
+    abspath = abspath.lexically_normal();
```
Add unit tests on this ? To make sure that old & new behavior are consistent

payetvin committed:
use throw and stl error msg

payetvin committed:
Merge branch 'fix/yuni-fs' of https://github.com/AntaresSimulatorTeam/Antares_Simulator into fix/yuni-fs

payetvin committed:
move readfile function to libs io

sylvlecl commented on src/libs/antares/io/include/antares/io/file.h:
```diff
@@ -21,14 +21,17 @@
 */
 bool IOFileSetContent(const AnyString& filename, const AnyString& content);
 
+std::string readFile(const std::filesystem::path&);
```
Let's wrap it into a namespace like `Antares::io`

payetvin committed:
add namespace Antares::IO

payetvin committed:
remove IO error

payetvin committed:
fix compile

payetvin committed:
unit test for yuni and std absolute

payetvin committed:
unit test for yuni and std normalize path

payetvin committed:
Merge branch 'develop' into fix/yuni-fs

payetvin committed:
try fix compile windows

payetvin committed:
Merge branch 'develop' into fix/yuni-fs

payetvin committed:
Merge branch 'develop' into fix/yuni-fs

payetvin committed:
add yuni logs overload for path

payetvin committed:
remove string conversion for logs

payetvin committed:
use implicit path conversion for lgos

payetvin committed:
fix compile

payetvin committed:
try revert

JasonMarechal25 commented on src/ext/yuni/src/yuni/core/string/traits/append.h:
```diff
@@ -294,6 +295,18 @@ class Append<CStringT, std::list<T>> final
     }
 };
 
+namespace fs = std::filesystem;
```
You're declaring fs but don't use it

JasonMarechal25 commented on src/libs/antares/io/file.cpp:
```diff
@@ -24,29 +24,53 @@
+    }
+
+    using Iterator = std::istreambuf_iterator<char>;
+    std::string content(Iterator{file}, Iterator{});
```
https://stackoverflow.com/a/2602060
If readfile is used to open large file it may good to improve the implementation.
An issue may be opened to talk about the subject

JasonMarechal25 commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -473,17 +475,16 @@ bool AreaLinksInternalLoadFromProperty(AreaLink& link, const String& key, const
-    if (!ini.open(buffer))
+    if (!ini.open(path.string()))
     {
         return 0;
```
```suggestion
        return false;
```

JasonMarechal25 commented on src/libs/antares/study/area/links.cpp:
```diff
@@ -496,7 +497,7 @@ bool AreaLinksLoadFromFolder(Study& study, AreaList* l, Area* area, const AnyStr
         // Getting the name of the area
-        buffer.clear();
+        String buffer;
         TransformNameIntoID(s->name, buffer);
```
We could make a new version of TransformNameIntoID that return a std::string and avoid the second parameter.

It would make some code cleaner.

JasonMarechal25 commented on src/libs/antares/study/header.cpp:
Do  we have unit tests for those function ?

If not would it be good to add some and ensure that we have the same behaviour before/after ?

JasonMarechal25 commented on src/libs/antares/study/header.cpp:
```diff
@@ -205,27 +205,25 @@ bool StudyHeader::saveToFile(const AnyString& filename, bool upgradeVersion)
 
-    // foldernormalization
-    String abspath, directory;
-    IO::MakeAbsolute(abspath, folder);
```
MakeAbsolute behaviour is weird due to is IsAbsolute
Windows style path are supported under linux contrary to fs::absolute.
On unix:
IO::IsAbsolute(C:/plop) => true
filesystem::path::is_absolute(C:/plop) => false

It sould not be an issue (or even supported), but well, it bares mentionning

JasonMarechal25 commented on src/libs/antares/study/header.cpp:
```diff
@@ -205,27 +205,25 @@ bool StudyHeader::saveToFile(const AnyString& filename, bool upgradeVersion)
+    abspath = abspath.lexically_normal();
 
-    if (not directory.empty() and IO::Directory::Exists(directory))
+    if (!abspath.empty() && fs::exists(abspath))
```
Is empty  test mendatory ? fs::exists("") should return false

JasonMarechal25 commented on src/libs/antares/study/include/antares/study/area/area.h:
```diff
@@ -34,6 +35,7 @@
 #include "antares/study/filter.h"
 #include "antares/study/parts/parts.h"
 
+#include "../parts/parts.h"
```
Please use full path

payetvin committed:
comments

guilpier-code commented on src/libs/antares/io/file.cpp:
```diff
@@ -24,29 +24,53 @@
+}
 
-bool IOFileSetContent(const AnyString& filename, const AnyString& content)
+bool fileSetContent(const std::string& filename, const std::string& content)
```
Fonction **fileSetContent** : 
it's the one and only function of this source file still containing references to **Yuni**.
It's uselessly complicated : 
- several attempts to open a file in writing mode 
- trying to identify the cause of the failure

Nobody does that.
It can be handled by STL well enough.
We should take example on function **readFile**, shorter and more clear (in the history of this source file, **readFile** came out much more recently than **fileSetContent**).
So we should rewrite this function with **std** things, like **readFile**.
It would be quick and easy to test, as **fileSetContent** is called twice in the whole code, and only from **immediate_file_writer.cpp**.
After that, no more reference to **Yuni** in this source file.

flomnes replied:
I agree. This was meant to allow the user to free up some space. But it only makes sense on a desktop environment

flomnes commented on src/ui/simulator/windows/sets.cpp:
```diff
@@ -581,7 +581,8 @@ void Sets::saveToStudy()
     // ... and reloading it
     // IO::File::LoadFromFile(study->simulation.comments, pTempFile);
-    study->setsOfAreas.loadFromFile(pTempFile);
+    std::filesystem::path p = std::filesystem::path(pTempFile.c_str());
```
```suggestion
    std::filesystem::path p(pTempFile.c_str());
```

JasonMarechal25 replied:
```suggestion
    std::filesystem::path p{pTempFile.c_str()};
```

flomnes commented on src/tests/src/libs/antares/test_utils.cpp:
```diff
@@ -83,3 +88,29 @@ BOOST_AUTO_TEST_CASE(test_transform_name_into_id)
+    Yuni::IO::Normalize(yuniNorm, p2.string());
+
+    BOOST_CHECK(p2.lexically_normal().string() == yuniNorm);
+}
```
Make two test-cases, don't mix up tests.

flomnes commented on src/solver/misc/options.cpp:
```diff
@@ -201,10 +198,10 @@ void checkAndCorrectSettingsAndOptions(Settings& settings, Data::StudyLoadOption
-        IO::File::Stream pidfile;
-        if (pidfile.openRW(optPID))
+        std::ofstream pidfile(optPID);
+        if (pidfile.is_open())
```
```suggestion
        if (std::ofstream pidfile(optPID); pidfile.is_open())
```

flomnes commented on src/tests/src/libs/antares/test_utils.cpp:
Add tests for `IsAbsolute`

flomnes commented:
`antares-solver` can't launch a simulation after this PR. Locally, I get
```
[2024-05-16 11:33:51][solver][fatal] The folder `/home/omnesflo/Téléchargements/valid-named-mps/binding-constraints/` does not seem to be a valid study
[2024-05-16 11:33:51][solver][fatal] Aborting now. (warning: no file log available)
[2024-05-16 11:33:51][solver][warns] No log file available
```

guilpier-code commented on src/libs/antares/writer/immediate_file_writer.cpp:
```diff
@@ -83,62 +39,67 @@ ImmediateFileResultWriter::ImmediateFileResultWriter(const char* folderOutput):
-static bool prepareDirectoryHierarchy(const YString& root,
-                                      const std::string& entryPath,
-                                      Yuni::String& output)
+static bool prepareDirectoryHierarchy(const fs::path& root,
```
I find this class difficult to understand.

- It seems it handles **results** (I understand **simulation results**) but it also handles logs (when it's called from **importlogs.cpp**)
- Method **addEntryFromFile** : every time I cross this method, I wonder what it does. I must read the body of this function to understand, and I find it hard to read. 
  - the function's name is unclear to me, 
  - what the arguments represent is mysterious to me, 
  - I don't understand why we make a file system copy.
- Function **prepareDirectoryHierarchy** has name unclear to me : actually it concatenates path elements and then create the resulting path on disk. We should remove this function and keep 2 functions, one for concatenation, the other for creation on disk. 
Besides, it has an output argument used to store the result of this function. This result should be returned instead.
- This class is supposed to flush something but it does not : the **flush** function is defined but empty. It looks like a problem of conception.

flomnes replied:
out of scope

payetvin committed:
Fix typo in if

flomnes committed:
Apply suggestions from code review

Co-authored-by: Jason Maréchal <45510813+JasonMarechal25@users.noreply.github.com>

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/yuni-fs

flomnes committed:
Split tests

flomnes committed:
Enrich & simplify tests

flomnes committed:
Remove useless test-case (is_absolute is not used for now)

payetvin committed:
Add transform name into id return

payetvin committed:
Merge branch 'fix/yuni-fs' of https://github.com/AntaresSimulatorTeam/Antares_Simulator into fix/yuni-fs

payetvin committed:
merge develop

payetvin committed:
Merge branch 'develop' into fix/yuni-fs

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/yuni-fs
